### PR TITLE
Rebuild project

### PR DESCRIPTION
**Pull request by anhnt97**
Rebuild project according to the MVC model

### CODE REVIEW

File: controllers/home_controller.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package controllers
```
Comment (hbl-duytv) [inline controllers/home_controller.go]: file name `home_controller.go` should be `home.go`

File: controllers/home_controller.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/hbl-duytv/intern-csm/models"
+	"github.com/hbl-duytv/intern-csm/services"
+)
+
+func GetAllEditorUser() []models.TransformUser {
+	user := []models.User{}
+	transformUSer := []models.TransformUser{}
```
Comment (hbl-duytv) [inline controllers/home_controller.go]: isn't needed

File: controllers/home_controller.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+	services.DB.Find(&user, "type=?", 0)
```
Comment (hbl-duytv) [inline controllers/home_controller.go]: move to `services` package

File: controllers/home_controller.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/hbl-duytv/intern-csm/models"
+	"github.com/hbl-duytv/intern-csm/services"
+)
+
+func GetAllEditorUser() []models.TransformUser {
+	user := []models.User{}
```
Comment (hbl-duytv) [inline controllers/home_controller.go]: use `var` to not create memory

File: controllers/home_controller.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+	if len(user) == 0 {
+		return nil
+	}
+	for _, v := range user {
+		transformUSer = append(transformUSer,
+			models.TransformUser{v.ID, v.Username, v.Email, v.Type, v.Gender, v.BirthDay, v.PhoneNumber, v.Status})
+	}
+	return transformUSer
+}
+func RenderHome(c *gin.Context) {
```
Comment (hbl-duytv) [inline controllers/home_controller.go]: `RenderHome(c)` should be `Home(c)`

File: controllers/login_controller.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/gin-gonic/contrib/sessions"
+	"github.com/gin-gonic/gin"
+	"github.com/hbl-duytv/intern-csm/models"
+	"github.com/hbl-duytv/intern-csm/services"
+)
+
+func AuthRequired() gin.HandlerFunc {
```
Comment (hbl-duytv) [inline controllers/login_controller.go]: move to `middleware` package

File: controllers/login_controller.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	return func(c *gin.Context) {
+		session := sessions.Default(c)
+		user := session.Get("user")
+		if user == nil {
+			// You'd normally redirect to login page
+			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid session token"})
```
Comment (hbl-duytv) [inline controllers/login_controller.go]: redirect to Login page if not login

File: controllers/login_controller.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+	// hash password to md5
+	passwordMD5 := GetMD5Hash(password)
+	user := models.User{}
+	services.DB.Where("username = ? AND password = ?", username, passwordMD5).Find(&user)
+	if user.ID != 0 {
+		session.Set("user", username)
+		err := session.Save()
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, gin.H{"status": http.StatusUnauthorized, "error": "Failed to generate session token"})
+		} else {
```
Comment (hbl-duytv) [inline controllers/login_controller.go]: use `return` instead of `else`

File: controllers/login_controller.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+			// c.JSON(http.StatusOK, gin.H{"status": http.StatusOK, "message": "Successfully authenticated user"})
+			if user.Status == 1 {
```
Comment (hbl-duytv) [inline controllers/login_controller.go]: use constant

File: controllers/login_controller.go [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+		c.JSON(http.StatusUnauthorized, gin.H{"status": http.StatusNotFound, "error": "Authentication failed"})
+	}
+
+}
+func RegisterSuccess(c *gin.Context) {
+	username := c.Param("username")
+	password := c.Param("password")
+	email := c.Param("email")
+	passwordMD5 := GetMD5Hash(password)
+	if username != "" && password != "" && email != "" {
```
Comment (hbl-duytv) [inline controllers/login_controller.go]: Check info `Register` and `Request` by token

File: controllers/login_controller.go [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+	}
+	c.JSON(http.StatusOK, gin.H{"status": http.StatusOK, "data": transformUSer})
+}
+func ConfirmUserAfterRegister(c *gin.Context) {
+	var user models.User
+	idUser := c.Param("id")
+	services.DB.First(&user, idUser)
+	status := 1
+	services.DB.Model(&user).Update("status", status)
+	services.DB.Save(&user)
```
Comment (hbl-duytv) [inline controllers/login_controller.go]: isn't needed. Use update and where.

File: controllers/login_controller.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+	if user == nil {
+		c.JSON(http.StatusBadRequest, gin.H{"status": http.StatusNotFound, "error": "Invalid session token"})
+	} else {
+		log.Println(user)
+		session.Delete("user")
+		session.Save()
+		c.JSON(http.StatusOK, gin.H{"status": http.StatusOK, "message": "Successfully logged out"})
+	}
+}
+func GetMD5Hash(text string) string {
```
Comment (hbl-duytv) [inline controllers/login_controller.go]: create and move to `helper` package

File: controllers/login_controller.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+	hasher := md5.New()
+	hasher.Write([]byte(text))
+	return hex.EncodeToString(hasher.Sum(nil))
+}
+func GetUserNotACtive() []models.TransformUser {
```
Comment (hbl-duytv) [inline controllers/login_controller.go]: fix name

File: views/html/create-post.html [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+        <li class="active"><a href="./create-post.html"><i class="fa fa-file"></i> <span>Tạo bài viết</span></a></li>
+        <li><a href="./post-management.html"><i class="fa fa-bars"></i> <span>Quản lý bài viết</span></a></li>
+        <li><a href="./editor-management.html"><i class="fa fa-users"></i> <span>Quản lý biên tập viên</span></a></li>
+      </ul>
+    </section>
+    <!-- /.sidebar -->
+  </aside>
+  <!-- =============================================== -->
+  <!-- Content Wrapper. Contains page content -->
+  <div class="content-wrapper">
```
Comment (hbl-duytv) [inline views/html/create-post.html:45]: use Go `template`

**Review by hbl-duytv — COMMENTED** [review]
@anhnt97 @trungduc08 Please fix comment!

File: controllers/login.go [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+	"strings"
+
+	"github.com/gin-gonic/contrib/sessions"
+	"github.com/gin-gonic/gin"
+	"github.com/hbl-duytv/intern-csm/helper"
+	"github.com/hbl-duytv/intern-csm/models"
+	"github.com/hbl-duytv/intern-csm/services"
+)
+
+const UserActived = 1
```
Comment (hbl-duytv) [inline controllers/login.go]: create and move to constant/ folder

File: controllers/login.go [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+	password := c.PostForm("password")
+
+	if strings.Trim(username, " ") == "" || strings.Trim(password, " ") == "" {
+		c.JSON(http.StatusUnauthorized, gin.H{"error": "Parameters can't be empty"})
+		return
+	}
+	// hash password to md5
+	passwordMD5 := helper.GetMD5Hash(password)
+	user := models.User{}
+	services.DB.Where("username = ? AND password = ?", username, passwordMD5).Find(&user)
```
Comment (hbl-duytv) [inline controllers/login.go]: move DB to service

File: controllers/login.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+		session.Set("user", username)
+		err := session.Save()
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, gin.H{"status": http.StatusUnauthorized, "error": "Failed to generate session token"})
+			return
+		}
+		if user.Status == UserActived {
+			c.Redirect(301, "/home")
+		} else {
+			message := []byte("Tài khoản chưa được kích hoạt, vui lòng đợi kích hoạt từ người quản trị!")
```
Comment (hbl-duytv) [inline controllers/login.go]: use constant

File: controllers/login.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+		c.JSON(http.StatusUnauthorized, gin.H{"status": http.StatusNotFound, "error": "Authentication failed"})
+	}
+
+}
+func RegisterSuccess(c *gin.Context) {
+	username := c.Param("username")
+	password := c.Param("password")
+	email := c.Param("email")
+	if username != "" && password != "" && email != "" {
+		services.CreateUser(username, password, email, 0, 0)
```
Comment (hbl-duytv) [inline controllers/login.go]: check returned error

File: controllers/login.go [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+	} else {
+		messageFail := []byte("Xác nhận tài khoản thất bại, vui lòng thử lại!")
+		c.Data(http.StatusOK, "text/html; charset=utf-8", messageFail)
+	}
+}
+func SendConfirmRegister(c *gin.Context) {
+	username := c.PostForm("username")
+	password := c.PostForm("password")
+	email := c.PostForm("email")
+	urlConfirm := "http://localhost:8000/confirm-register/" + username + "/" + password + "/" + email
```
Comment (hbl-duytv) [inline controllers/login.go]: use token

File: controllers/login.go [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+	if username != "" && password != "" && email != "" {
+		services.CreateUser(username, password, email, 0, 0)
+		messageSuccess := []byte("Xác nhận tài khoản thành công, vui lòng đợi kích hoạt từ người quản trị!")
+		c.Data(http.StatusOK, "text/html; charset=utf-8", messageSuccess)
+	} else {
+		messageFail := []byte("Xác nhận tài khoản thất bại, vui lòng thử lại!")
+		c.Data(http.StatusOK, "text/html; charset=utf-8", messageFail)
+	}
+}
+func SendConfirmRegister(c *gin.Context) {
```
Comment (hbl-duytv) [inline controllers/login.go:64]: combine to func RegisterSuccess()

File: controllers/login.go [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+	email := c.PostForm("email")
+	urlConfirm := "http://localhost:8000/confirm-register/" + username + "/" + password + "/" + email
+	massageEmailConfirm := "<div>Bạn đã đăng ký tài khoản biên tập viên, vui lòng xác nhận :</div><button><a href=\"" + urlConfirm + "\">Xác nhận đăng ký</a></button>"
+	SendMail(email, massageEmailConfirm)
+	c.Data(http.StatusOK, "text/html; charset=utf-8", []byte("Gửi mail xác nhận thành công, vui lòng check mail để xác nhận đăng ký tài khoản!"))
+}
+func ConfirmUserAfterRegister(c *gin.Context) {
+	idUser := c.Param("id")
+	user := services.GetUserByID(idUser)
+	services.UpdateStatusUser(1, &user)
```
Comment (hbl-duytv) [inline controllers/login.go]: use constant

File: controllers/login.go [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+		c.Redirect(301, "login")
+		return
+	}
+	session.Delete("user")
+	session.Save()
+	c.Redirect(301, "login")
+}
+func CheckUserExist(c *gin.Context) {
+	username := c.PostForm("username")
+	user := services.GetUserByID(username)
```
Comment (hbl-duytv) [inline controllers/login.go]: username = userID?

File: controllers/sendmail.go [hunk 23]
```diff
@@ -0,0 +1,20 @@
+package controllers
+
+import (
+	"crypto/tls"
+
+	gomail "gopkg.in/gomail.v2"
+)
+
+func SendMail(emailReceiver string, message string) {
```
Comment (hbl-duytv) [inline controllers/sendmail.go]: should be in services/

File: middleware/auth.go [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+}
+func AuthAdminRequired() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		session := sessions.Default(c)
+		username := session.Get("user")
+		if username == nil {
+			c.Redirect(301, "login")
+		} else if usernameString, ok := username.(string); ok {
+			user := services.GetUserByUsername(usernameString)
+			if user.Type == 1 {
```
Comment (hbl-duytv) [inline middleware/auth.go]: use constant

File: services/user.go [hunk 25]
```diff
@@ -0,0 +1,45 @@
+package services
+
+import (
+	"github.com/hbl-duytv/intern-csm/helper"
+	"github.com/hbl-duytv/intern-csm/models"
+)
+
+func GetAllEditorUser() []models.User {
+	var users []models.User
+	DB.Find(&users, "type=?", 0)
```
Comment (hbl-duytv) [inline services/user.go]: return error

**Review by hbl-duytv — COMMENTED** [review]
@trungduc08 . Please fix!

File: controllers/editor.go [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+	idUser, _ := strconv.Atoi(c.PostForm("id"))
+	if services.UpdateStatusUser(idUser, constant.DEACTIVE_NUMBER) == nil {
+		c.JSON(http.StatusOK, gin.H{"status": http.StatusOK, "message": "Updated status user successfully!"})
+	} else {
+		c.JSON(http.StatusNotFound, gin.H{"status": http.StatusNotFound, "message": "Updated status user fail!"})
+	}
+}
+func DeleteUser(c *gin.Context) {
+	idUser, _ := strconv.Atoi(c.PostForm("id"))
+	if services.DeletePost(idUser) == nil {
```
Comment (hbl-duytv) [inline controllers/editor.go]: `DeleteUser`

File: controllers/editor.go [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+	name := c.PostForm("name")
+	gender := c.PostForm("gender")
+	birthday := c.PostForm("birthday")
+	token := helper.GetToken(username)
+	if phoneNumber, error := strconv.Atoi(c.PostForm("phone_number")); error != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"status": http.StatusBadRequest, "message": "Phone number error!"})
+	} else if status, error := strconv.Atoi(c.PostForm("status")); error != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"status": http.StatusBadRequest, "message": "Status error!"})
+	} else {
+		if services.CreateAccount(username, password, email, name, gender, birthday, phoneNumber, status, constant.DEACTIVE_NUMBER, token, constant.DEACTIVE_NUMBER) == nil {
```
Comment (hbl-duytv) [inline controllers/editor.go]: pass `User{}` to function

File: constant/const.go [hunk 28]
```diff
@@ -0,0 +1,5 @@
+package constant
+
+const ACTIVE_NUMBER int = 1
```
Comment (hbl-duytv) [inline constant/const.go]: use PascalCase for constant

File: controllers/editor.go [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/hbl-duytv/intern-csm/helper"
+	"github.com/hbl-duytv/intern-csm/services"
+)
+
+func ActiveEditorUser(c *gin.Context) {
+
+	idUser, _ := strconv.Atoi(c.PostForm("id"))
+	if services.UpdateStatusUser(idUser, constant.ACTIVE_NUMBER) == nil {
+		c.JSON(http.StatusOK, gin.H{"status": http.StatusOK, "message": "Updated status user successfully!"})
+	} else {
```
Comment (hbl-duytv) [inline controllers/editor.go]: remove `else`

File: controllers/login.go [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+	} else {
+		c.JSON(http.StatusUnauthorized, gin.H{"status": http.StatusNotFound, "error": "Authentication failed"})
+	}
+
+}
+func RegisterSuccess(c *gin.Context) {
+	var messageSuccess []byte
+	token := c.Param("token")
+	if user, err := services.GetUserByToken(token); err == nil {
+		// fmt.Println("inf: ", user.Username, user.Confirm)
```
Comment (hbl-duytv) [inline controllers/login.go]: remove comment

File: controllers/posts.go [hunk 31]
```diff
@@ -0,0 +1,177 @@
...
+
+}
+func ActiveStatusPost(c *gin.Context) {
+	session := sessions.Default(c)
+	username := session.Get("user")
+	var user models.User
+	if usernameString, ok := username.(string); ok {
+		user, _ = services.GetUserByUsername(usernameString)
+	} else {
+		c.Redirect(constant.DIRECT_STATUS, "/login")
```
Comment (hbl-duytv) [inline controllers/posts.go]: move to middleware to check login

File: middleware/auth.go [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+}
+func AuthAdminRequired() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		session := sessions.Default(c)
+		username := session.Get("user")
+		if username == nil {
+			c.Redirect(301, "login")
+		} else if usernameString, ok := username.(string); ok {
+			user, _ := services.GetUserByUsername(usernameString)
+			if user.Type == constant.ACTIVE_NUMBER {
```
Comment (hbl-duytv) [inline middleware/auth.go]: should be TypeAdmin

File: models/post.go [hunk 33]
```diff
@@ -0,0 +1,26 @@
...
+	Creator     int       `json:"creator"`
+	Title       string    `json:"title"`
+	Topic       string    `json:"topic"`
+	Description string    `json:"description"`
+	Content     string    `json:"content"`
+	Status      int       `json:"status"`
+	CreatedAt   time.Time `json:"created_at"`
+	UpdatedAt   time.Time `json:"updated_at"`
+}
+type TransformPost struct {
```
Comment (hbl-duytv) [inline models/post.go:16]: remove `TransformPost`

File: services/post.go [hunk 34]
```diff
@@ -0,0 +1,117 @@
...
+)
+
+func GetPostWithAdminPermission() ([]models.TransformPost, error) {
+	var posts []models.Post
+	var transformPosts []models.TransformPost
+	if err := DB.Find(&posts).Error; err != nil {
+		return transformPosts, err
+	}
+	for _, v := range posts {
+		if user, err := GetUserByID((v.Creator)); err == nil {
```
Comment (hbl-duytv) [inline services/post.go]: no query in loop

File: services/user.go [hunk 35]
```diff
@@ -0,0 +1,130 @@
...
+	return nil
+}
+func GetTimeCreateUSer(id int) (int, int, error) {
+	var user models.User
+	if err := DB.Where("id=?", id).Find(&user).Error; err != nil {
+		return int(user.CreatedAt.Month()), int(user.CreatedAt.Year()), err
+	}
+	return int(user.CreatedAt.Month()), int(user.CreatedAt.Year()), nil
+}
+func CheckTimeToConfirmUser(user models.User) bool {
```
Comment (hbl-duytv) [inline services/user.go]: Comment to describe function

**Review by hbl-duytv — COMMENTED** [review]
@anhnt97 @trungduc08 Please fix

File: models/post.go (hunk 33 above)
Comment (trungduc08) [reply models/post.go:16]: create TransformPost to get name of creator because Post must have creator id

File: constant/const.go [hunk 36]
```diff
@@ -0,0 +1,7 @@
+package constant
+
+const ActiveNumber int = 1
+const DeactiveNumber int = 0
+const DirectStatus int = 301
```
Comment (hbl-duytv) [inline constant/const.go:5]: use `http.StatusMovedPermanently`

File: controllers/editor.go [hunk 37]
```diff
@@ -0,0 +1,91 @@
...
+		c.JSON(http.StatusBadRequest, gin.H{"status": http.StatusBadRequest, "message": "User create success!"})
+
+	}
+}
+
+func RenderEditorManagement(c *gin.Context) {
+
+	editors, _ := services.GetAllEditorUser()
+	session := sessions.Default(c)
+	username := session.Get("user")
```
Comment (hbl-duytv) [inline controllers/editor.go]: create func `GetCurrentUser(c)`

File: controllers/login.go [hunk 38]
```diff
@@ -0,0 +1,143 @@
...
+		return
+	}
+	session.Delete("user")
+	session.Save()
+	c.Redirect(constant.DirectStatus, "login")
+}
+func CheckUserExist(c *gin.Context) {
+	username := c.PostForm("username")
+	if user, err := services.GetUserByUsername(username); err == nil {
+		if user.ID == constant.DeactiveNumber {
```
Comment (hbl-duytv) [inline controllers/login.go]: remove

File: controllers/login.go [hunk 39]
```diff
@@ -0,0 +1,143 @@
...
+		}
+	}
+
+	c.JSON(http.StatusOK, gin.H{"check": false, "message": "User exist!"})
+	return
+}
+func CheckEmailExist(c *gin.Context) {
+	email := c.PostForm("email")
+	if user, err := services.GetUserByEmail(email); err == nil {
+		if user.ID == constant.DeactiveNumber {
```
Comment (hbl-duytv) [inline controllers/login.go]: remove

File: controllers/posts.go [hunk 40]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/hbl-duytv/intern-csm/services"
+)
+
+func RenderPostManagementAdmin(c *gin.Context) {
+	session := sessions.Default(c)
+	username := session.Get("user")
+	if usernameString, ok := username.(string); ok {
+		if user, err := services.GetUserByUsername(usernameString); err == nil {
+			if posts, err := services.GetPostWithAdminPermission(); err == nil {
+				month, year, _ := services.GetTimeCreateUSer(user.ID)
```
Comment (hbl-duytv) [inline controllers/posts.go]: handle error

File: controllers/posts.go [hunk 41]
```diff
@@ -0,0 +1,163 @@
...
+		c.HTML(http.StatusOK, "master.html", gin.H{"month": month, "year": year, "post": post, "index": 3, "title": "Update Post"})
+		return
+	}
+	c.Redirect(constant.DirectStatus, "/home")
+}
+
+func RenderDetailPost(c *gin.Context) {
+	idPost, _ := strconv.Atoi(c.Param("id"))
+	var usernames []string
+	post, _ := services.GetPostById(idPost)
```
Comment (hbl-duytv) [inline controllers/posts.go]: handle error

File: controllers/posts.go [hunk 42]
```diff
@@ -0,0 +1,163 @@
...
+	title := c.PostForm("title")
+	topic := c.PostForm("topic")
+	description := c.PostForm("description")
+	content := c.PostForm("content")
+	fmt.Println(title, topic, description, content)
+	if strings.Trim(title, " ") == "" || strings.Trim(topic, " ") == "" || description == "" || strings.Trim(content, " ") == "" {
+		c.JSON(http.StatusBadRequest, gin.H{"status": http.StatusBadRequest, "error": "Nhập đầy đủ thông tin"})
+		return
+	}
+	if services.CreatePost(creator, title, topic, description, content) == nil {
```
Comment (hbl-duytv) [inline controllers/posts.go]: use model `Post`

File: services/post.go [hunk 43]
```diff
@@ -0,0 +1,125 @@
...
+
+import (
+	"time"
+
+	"github.com/hbl-duytv/intern-csm/models"
+)
+
+func GetPostWithAdminPermission() ([]models.TransformPost, error) {
+	var posts []models.TransformPost
+	rows, err := DB.Table("post").Select("post.id, user.name, post.title, post.topic, post.description, post.content, post.status, post.created_at, post.updated_at").Joins("join user on user.id = post.creator_id").Rows()
```
Comment (hbl-duytv) [inline services/post.go]: use alias `as`

File: services/post.go [hunk 44]
```diff
@@ -0,0 +1,125 @@
...
+	return nil
+}
+func CreatePost(idCreator int, title, topic, des, content string) error {
+	newPost := models.Post{
+		Creator:     idCreator,
+		Title:       title,
+		Topic:       topic,
+		Description: des,
+		Content:     content,
+		Status:      0,
```
Comment (hbl-duytv) [inline services/post.go]: use constant

**Review by hbl-duytv — COMMENTED** [review]
@anhnt97 @trungduc08 Please fix

**Review by hbl-duytv — APPROVED** [review]

**Pull request merged**